### ISSUE TITLE
SameDiff: softmax, log softmax (and derivatives) should have dimension arg (or args?)

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/nd4j#2541]
https://github.com/deeplearning4j/nd4j/blob/86de11c3a44401a5a5afa70e43c0591b18bbcb33/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java#L2035-L2051

https://github.com/deeplearning4j/nd4j/blob/86de11c3a44401a5a5afa70e43c0591b18bbcb33/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/functions/DifferentialFunctionFactory.java#L401

https://github.com/deeplearning4j/nd4j/blob/753f3c0e4545714f47106db4e88d1900540bb350/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/SoftMax.java#L41-L53

Compare:
https://www.tensorflow.org/api_docs/python/tf/nn/softmax

**Comment by raver119** [comment]
side note: softmax, logsoftmax etc will be removed from legacy ops, and custom ops will be used instead.
